Dynamically Linked Library in Presto CPP (#24330)
prestodb/presto: presto-docs/src/main/sphinx/presto_cpp/plugin/function_plugin.rst

=======================
Function Plugin
=======================

Creating a Shared Library for UDFs
----------------------------------

> 

Reviewer: For people new to Presto, consider adding an introductory sentence, maybe a paragraph, to expand the acronym UDF, explain what it is, and what they are good for. You don't need much, just a little to help people not as familiar with the topic.

> 1. To create the UDF, create a new C++ file which follows the following format:

Reviewer: ```suggestion
1. To create the UDF, create a new C++ file in the following format:
```
Minor suggestion for conciseness.

.. code-block:: c++

   #include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"

   template <typename T>
   struct nameOfStruct {
      FOLLY_ALWAYS_INLINE bool call(int64_t& result) {
         ...
      }
   };

   extern "C" {
      void registry() {
         facebook::presto::registerPrestoFunction<
            nameOfStruct,
            int64_t>("function_name");
         }
   }  

> Note: the int64_t return type can be changed as needed. For more examples, See the `README <https://github.com/prestodb/presto-native-execution/main/dynamic_registry/README.md>`_ 

Reviewer: ```suggestion
Note: the ``int64_t`` return type can be changed as needed. For more examples, see the `README <https://github.com/prestodb/presto-native-execution/main/dynamic_registry/README.md>`_.
```

2. Create a shared library which may be made using CMakeLists like the following:

.. code-block:: text

   add_library(name_of_dynamic_fn SHARED TestFunction.cpp)
   target_link_libraries(name_of_dynamic_fn PRIVATE fmt::fmt Folly::folly gflags::gflags)

> 3. `Drop your shared libraries into the plugin directory <../plugin.rst>`_

Reviewer: ```suggestion
3. Put your shared libraries in the plugin directory described in :doc:`../plugin`. 
```

I tested this link format in a local doc build, and it worked. Minor revision to the text suggested.

> =======================

Reviewer 2: Do we still need the README if we have the documentation here?
There is some divergence between the readme and this.

Perhaps the readme can just point to the documentation link instead? We've done something similar with the velox part of the PR.

We could also link to the Velox piece that is underlying the Presto part.

Reviewer 3: Can we add some documentation on how to develop these plugins, with examples?

Author: hey @tdcmeehan could you help me understand your ask a little better? 

Currently, I'm providing the boilerplate portions for the json config and UDF creating bit along with the CMakeLists as an option to compile the shared library UDFs. 

additionally, specific examples are linked in the doc that points to the examples directory that covers various use cases. 

> see the `examples <https://github.com/prestodb/presto/tree/master/presto-native-execution/main/dynamic_registry/examples>`_.

Was there anything (else) youd like seen in this doc?

Function Plugin
=======================

Creating a Shared Library for UDFs
----------------------------------

> User defined functions (UDFs) enhance Presto CPP's capabilities by making it easy to create custom functions without the need to rebuild the executible. 

Reviewer: ```suggestion
User defined functions (UDFs) allow users to create custom functions without the need to rebuild the executable. 
```
Minor revision to emphasize that users can create UDFs. It's obvious, or should be, but emphasizing it helps. Removing "enhance Presto CPP's capabilities" focuses the sentence on what is a UDF.

Author: i see the difference between the two versions and agree! Thank you

> There are many benefits to UDFs other than shorter compile times, such as:

Reviewer 2: "other than shorter compile times". An user wouldn't care about compile times. I would just leave the "other than ..." out and simply state the benefits.

> There are many benefits to UDFs beyond the shorter (re)compile times including but not limited to:

Reviewer: ```suggestion
There are many benefits to UDFs other than shorter compile times, such as:
```
Minor revision suggested for conciseness.

* Simplify SQL queries by creating UDFs for repetitive logic.
* Implement custom logic pertaining to the specific business use cases of the users.

> * Once defined, easily reusible and called multiple times just like built in functions.

Reviewer: ```suggestion
* Once defined, easily reusable and called multiple times just like built in functions.
```

Author: addressed, great catch! i didnt notice it earlier too.

1. To create the UDF, create a new C++ file in the following format:

   .. code-block:: c++

      #include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"

      template <typename T>

>       struct nameOfStruct {

Reviewer 2: In most (all?) the code conventions a struct/class name starts with upper case for a class or struct so should do the same here (and then usage later).

         FOLLY_ALWAYS_INLINE bool call(int64_t& result) {
            ...
         }
      };

      extern "C" {
         void registry() {
            facebook::presto::registerPrestoFunction<
               nameOfStruct,
               int64_t>("function_name");
            }
      }  

   Note: the ``int64_t`` return type can be changed as needed. For more examples, see the `README <https://github.com/prestodb/presto-native-execution/main/dynamic_registry/README.md>`_.

> 2. Create a shared library which may be made using CMakeLists like the following:

Reviewer 2: Nit: ... may be made using a ``CMakeLists.txt`` like  ...

We should highlight the file name because it is fixed for a directory.

   .. code-block:: text

      add_library(name_of_dynamic_fn SHARED TestFunction.cpp)
      target_link_libraries(name_of_dynamic_fn PRIVATE fmt::fmt Folly::folly gflags::gflags)

> 3. Put your shared libraries in the plugin directory described in :doc:`../plugin`.

Reviewer 2: This is not true anymore.
The user must configure the new `plugin.dir` variable to indicate where to pick up the new libraries. If it is the default (empty string) nothing happens.
On the CPP side this is good because we likely need some kind of option where libraries are not loaded even if a library is present - aka never load anything dynamically as otherwise it is too easy to exploit. Java is a bit more secure in that regard.

Author: I see how this was worded in a confusing way. My intent here was to point out that you have to go through the setup steps in the plugin.rst file which has the "configure the new plugin.dir variable" step there. I tried to rewrite it to make it more obvious, please let me know if that addresses your concerns @czentgr

=======================
Function Plugin
=======================

Creating a Shared Library for UDFs
----------------------------------
User defined functions (UDFs) allow users to create custom functions without the need to rebuild the executable. 
There are many benefits to UDFs, such as:

* Simplify SQL queries by creating UDFs for repetitive logic.
* Implement custom logic pertaining to the specific business use cases of the users.
* Once defined, easily reusable and called multiple times just like built in functions.
* Shorter compile times.

> 1. To create the UDF, create a new C++ file in the following format:

Reviewer 4: I think it is better to give a file name here, that will better align wiht the following content. For exmaple TestFunction.cpp is ok.

   .. code-block:: c++

      #include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"

      template <typename T>
      struct NameOfStruct {
         FOLLY_ALWAYS_INLINE bool call(int64_t& result) {
            ...
         }
      };

      extern "C" {
         void registry() {
            facebook::presto::registerPrestoFunction<

>                nameOfStruct,

Reviewer 4: nit: nameOfStruct to NameOfStruct.

===============
Function Plugin
===============

Creating a Shared Library for UDFs
----------------------------------
User defined functions (UDFs) allow users to create custom functions without the need to rebuild the executable. 
There are many benefits to UDFs, such as:

* Simplify SQL queries by creating UDFs for repetitive logic.
* Implement custom logic pertaining to the specific business use cases of the users.
* Once defined, easily reusable and called multiple times just like built in functions.
* Shorter compile times.

> 1. To create the UDF, create a new C++ file in the same format as the below file named ``ExampleFunction.cpp``:

Author: we want the shared libraries to be placed in the plugin directory, not necessarily the cpp file so i kept the plugin directory part out for now... pls lmk if that looks good!

> 1. To create the UDF, create a new C++ file in the same format as the below file named ``ExampleFunction.cpp``:

Reviewer: > we want the shared libraries to be placed in the plugin directory, not necessarily the cpp file so i kept the plugin directory part out for now... pls lmk if that looks good!

I understand what you write here, but it's in the current branch - screenshot of a local doc build from the updated branch. 

If you change "plugin directory" to tell the reader where you want them to create the .cpp file, I think that's the last thing in the doc. Thanks for being patient!

<img width="773" alt="Screenshot 2025-06-04 at 3 22 05 PM" src="https://github.com/user-attachments/assets/a8991d7b-6332-4fd0-98cb-e33b4a8a35ff" />

> 1. To create the UDF, create a new C++ file in the same format as the below file named ``ExampleFunction.cpp``:

Reviewer: @soumiiow, 
1) No worries! It's there (or not there) now, as you intended.
2) Thanks for the explanation! I understand now, it makes sense the way you have it.

            ...
         }
      };

      extern "C" {
         void registerExtensions() {
            facebook::presto::registerPrestoFunction<
               NameOfStruct,
               int64_t, facebook::velox::Varchar>("function_name", "test.namespace");
            }
      }  

   Note: The ``int64_t`` return type and the ``Varchar`` input type can be changed as needed. Additional or no arguments may be specified as well. For more examples, see the `examples <https://github.com/prestodb/presto/tree/master/presto-native-execution/main/dynamic_registry/examples>`_.
   Examples of typenames for return types and argument types include: 
      * simple types: BOOLEAN, TINYINT, SMALLINT, INTEGER, BIGINT, DOUBLE, TIMESTAMP, DATE, VARCHAR, VARBINARY.
      * complex types: ARRAY, MAP, ROW. They can be mixed with other simple and complex types. For example: MAP(INTEGER, ARRAY(BIGINT)).

2. Create a shared library which may be made using ``CMakeLists.txt`` like the following:

   .. code-block:: text

      add_library(name_of_dynamic_fn SHARED ExampleFunction.cpp)
      target_link_libraries(name_of_dynamic_fn PRIVATE fmt::fmt Folly::folly gflags::gflags)

> 3. Place your shared libraries in the plugin directory. The path to this directory must be inside ``presto-cpp/main/dynamic_registry/plugin``.

Reviewer: ```suggestion
3. Place your shared libraries in the ``presto-cpp/main/dynamic_registry/plugin`` directory. 
```

===============
Function Plugin
===============

Creating a Shared Library for UDFs
----------------------------------
User defined functions (UDFs) allow users to create custom functions without the need to rebuild the executable. 
There are many benefits to UDFs, such as:

* Simplify SQL queries by creating UDFs for repetitive logic.
* Implement custom logic pertaining to the specific business use cases of the users.
* Once defined, easily reusable and called multiple times just like built in functions.
* Shorter compile times.

> 1. To create the UDF, create a new C++ file named ``ExampleFunction.cpp`` in the same format as this example:

Author: good catch! thank you!

   .. code-block:: c++

      #include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"

      template <typename T>
      struct NameOfStruct {
         VELOX_DEFINE_FUNCTION_TYPES(T);
         FOLLY_ALWAYS_INLINE bool call(out_type<int64_t>& result, const arg_type<Varchar> in) {
            ...
         }
      };

      extern "C" {

>          void registerExtensions() {

Reviewer 2: Should we mention it that this name is required in the note below? It is mentioned in the examples.
We could say something like:

> The function ``registerExtensions`` is the entry point to execute the registration of the UDF and cannot be changed.

Author: i like this, ill add this as a comment for all example files!

            facebook::presto::registerPrestoFunction<
               NameOfStruct,
               int64_t, facebook::velox::Varchar>("function_name", "test.namespace");
            }
      }  

   Note: The ``int64_t`` return type and the ``Varchar`` input type can be changed as needed. Additional or no arguments may be specified as well. For more examples, see the `examples <https://github.com/soumiiow/presto/tree/dylib_new/presto-native-execution/presto_cpp/main/dynamic_registry>`_.
   Examples of typenames for return types and argument types include: 

   * simple types: BOOLEAN, TINYINT, SMALLINT, INTEGER, BIGINT, DOUBLE, TIMESTAMP, DATE, VARCHAR, VARBINARY.
   * complex types: ARRAY, MAP, ROW. They can be mixed with other simple and complex types. For example: MAP(INTEGER, ARRAY(BIGINT)).

2. Create a shared library using ``CMakeLists.txt`` like the following:

   .. code-block:: text

      add_library(name_of_dynamic_fn SHARED ExampleFunction.cpp)
      target_link_libraries(name_of_dynamic_fn PRIVATE fmt::fmt Folly::folly gflags::gflags)

> 3. Place your shared libraries in the ``presto-cpp/main/dynamic_registry/plugin`` directory. 

Author: good point. I'm unsure on what this path should be. let me paste the offline convo here as well. 

> i added a macro in presto-native-execution/presto_cpp/main/CMakeLists.txt for a TBD on location plugin directory. thoughts on that? and thoughts on plugin to be in dynamic_registry/plugin or presto-native-execution/plugin to be consistent with the docs below

It makes sense to keep it consistent with the spi docs. happy to make the change

There are many benefits to UDFs, such as:

* Simplify SQL queries by creating UDFs for repetitive logic.
* Implement custom logic pertaining to the specific business use cases of the users.
* Once defined, easily reusable and called multiple times just like built in functions.
* Shorter compile times.

1. To create the UDF, create a new C++ file in the same format as ``TestFunction.cpp`` below:

   .. code-block:: c++

      #include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"

      template <typename T>
      struct NameOfStruct {
         FOLLY_ALWAYS_INLINE bool call(int64_t& result) {
            ...
         }
      };

      extern "C" {
         void registryTest() {
            facebook::presto::registerPrestoFunction<
               NameOfStruct,
               int64_t>("function_name", "test.namespace");
            }
      }  

   Note: the ``int64_t`` return type, ``registryTest`` registry symbol name can be changed as needed. For more examples, see the `examples <https://github.com/prestodb/presto/tree/master/presto-native-execution/main/dynamic_registry/examples>`_.

> 2. Create a shared library which may be made using CMakeLists.txt like the following:

Reviewer: ```suggestion
2. Create a shared library which may be made using ``CMakeLists.txt`` like the following:
```

   .. code-block:: text

      add_library(name_of_dynamic_fn SHARED TestFunction.cpp)
      target_link_libraries(name_of_dynamic_fn PRIVATE fmt::fmt Folly::folly gflags::gflags)

3. Place your shared libraries in the plugin directory. The path to this directory needs to be the same as ``plugin.dir`` property set in :doc:`../plugin`.

> 4. Create a Json configuration file in the same format as below:

Reviewer: When I read this, I have the same question as my comment on line 18 - where should I create this file? Is there a name and example that could be included in the doc?

Reviewer: Same answer as above: let's not.

> 4. Create a Json configuration file in the same format as below:

Reviewer: ```suggestion
4. Create a JSON configuration file in the same format as below:
```

               "paramTypes": [
                  "INTEGER"
               ],
               "nameSpace": "presto.default",
               "routineCharacteristics": {
                  "language": "CPP",
                  "determinism": "DETERMINISTIC",
                  "nullCallClause": "CALLED_ON_NULL_INPUT"
               }
            }
            ]
         }
      }
      } 

   * ``dynamicLibrariesUdfMap``: required constant top-level field.
   * ``name_of_subdirectory``: replace this with the subdirectory your shared library is situated in inside the plugin directory.
   * ``function_name``: needs to be the same as the function name in ``TestFunction.cpp``.
   * ``outputType`` and  ``paramTypes``: must match the output and parameter types in ``TestFunction.cpp``.
   * ``fileName``: matches the name of the shared object (without a .so/.dylib suffix).
   * ``nameSpace``: optional field. omitting this field gives the function the default namespace similarly to the function registeration. must match the namespace of the function.  
   * ``docString``, ``routineCharacteristics``, ``functionKind``: collected for checking metadata.

> 5. Set the ``plugin.config`` property to the path of Json config. Instructions to set the property in :doc:`../plugin`.

Reviewer: ```suggestion
5. Set the ``plugin.config`` property to the path of the Json configuration file created in Step 4. See :doc:`../plugin`.
```

=======================
Function Plugin
=======================

Creating a Shared Library for UDFs
----------------------------------
User defined functions (UDFs) allow users to create custom functions without the need to rebuild the executable. 
There are many benefits to UDFs, such as:

* Simplify SQL queries by creating UDFs for repetitive logic.
* Implement custom logic pertaining to the specific business use cases of the users.
* Once defined, easily reusable and called multiple times just like built in functions.
* Shorter compile times.

1. To create the UDF, create a new C++ file in the same format as ``TestFunction.cpp`` below:

   .. code-block:: c++

      #include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"

      template <typename T>
      struct NameOfStruct {
         FOLLY_ALWAYS_INLINE bool call(int64_t& result) {
            ...
         }
      };

      extern "C" {
         void registryTest() {
            facebook::presto::registerPrestoFunction<
               NameOfStruct,
               int64_t>("function_name", "test.namespace");
            }
      }  

>    Note: the ``int64_t`` return type, ``registryTest`` registry symbol name can be changed as needed. For more examples, see the `examples <https://github.com/prestodb/presto/tree/master/presto-native-execution/main/dynamic_registry/examples>`_.

Reviewer: ```suggestion
   Note: The ``int64_t`` return type and the ``registryTest`` registry symbol name can be changed as needed. For more examples, see the `examples <https://github.com/prestodb/presto/tree/master/presto-native-execution/main/dynamic_registry/examples>`_.
```
Minor phrasing suggestion. 
As a second point, I don't find the linked examples either at the current destination, or this PR. If the examples don't exist, suggest removing this sentence or adding them (back in?) as part of this PR.

3. Place your shared libraries in the plugin directory. The path to this directory must be the same as the ``plugin.dir`` property set in :doc:`../plugin`.

4. Create a JSON configuration file in the same format as below:

   .. code-block::json  

      {
      "dynamicUdfSignatureMap": {
         "name_of_subdirectory": {
            "function_name": [
            {
               "docString": "",
               "functionKind": "SCALAR",
               "outputType": "INTEGER",
               "entrypoint": "registryTest", 
               "fileName": "name_of_dynamic_fn",
               "paramTypes": [
                  "INTEGER"
               ],
               "nameSpace": "presto.default",
               "routineCharacteristics": {
                  "language": "CPP",
                  "determinism": "DETERMINISTIC",
                  "nullCallClause": "CALLED_ON_NULL_INPUT"
               }
            }
            ]
         }
      }
      } 

   * ``dynamicUdfSignatureMap``: required constant top-level field.
   * ``name_of_subdirectory``: replace this with the subdirectory your shared library is situated in inside the plugin directory.

>    * ``function_name``: needs to be the same as the function name in ``TestFunction.cpp``.

Reviewer: ```suggestion
   * ``function_name``: needs to be the same as the function name in the file you created in Step 1.
```
Rephrasing suggested because you did not tell the reader to name the file they created ``TestFunction.cpp``.
Alternatively, recommend a name in Step 1 - maybe something like "For these instructions, the file you create is referred to as ``ExampleFunction.cpp``.", then use that name here and in the next line.

>    * ``outputType`` and  ``paramTypes``: must match the output and parameter types in ``TestFunction.cpp``. 

Reviewer: ```suggestion
   * ``outputType`` and  ``paramTypes``: must match the output and parameter types as in the file you created in Step 1. 
```
Rephrasing suggested because you did not tell the reader to name the file they created ``TestFunction.cpp``.

>       * Examples of typenames include: 

Reviewer: ```suggestion
      Examples of typenames include: 
```
You don't need this sentence as a single-item list item, with a sub-list underneath it. Make sure this line is indented to be inside the list item in line 79. 

Here's a screenshot of how it looks now: 
<img width="753" alt="Screenshot 2025-05-30 at 3 15 30 PM" src="https://github.com/user-attachments/assets/daf63c6b-f119-43d0-872a-757211499bac" />

Here's a screenshot of what I'm suggesting. I've included the next item for context of the layout: 

<img width="694" alt="Screenshot 2025-05-30 at 3 17 01 PM" src="https://github.com/user-attachments/assets/169631a8-51ac-4b73-babf-e453074f32dc" />

In my example I didn't put "Examples of typenames" to the left margin of the list item, but not for any specific reason, Tf you think "Examples of typenames"  looks better as a new line please do it that way.

         * simple types: BOOLEAN, TINYINT, SMALLINT, INTEGER, BIGINT, DOUBLE, TIMESTAMP, DATE, VARCHAR, VARBINARY.
         * complex types: ARRAY, MAP, ROW. They can be mixed with other simple and complex types. For example: MAP(INTEGER, ARRAY(BIGINT)).
   * ``registryTest``: needs to match the name of the registry function name in ``TestFunction.cpp``. If this field is not specified in the JSON config, it defaults to ``registerExtensions``.

>    * ``fileName``: matches the name of the shared object library (without a .so/.dylib suffix).

Reviewer: ```suggestion
   * ``fileName``: matches the name of the shared object library, without a ``.so`` or ``.dylib`` suffix.
```
Try to avoid ``/`` in documentation when you can. I've made one suggestion here, but if you think of a better one please use it!

>    * ``nameSpace``: optional field. omitting this field gives the function the default namespace similarly to the function registeration. must match the namespace of the function.  

Reviewer: ```suggestion
   * ``nameSpace``: optional field. Omitting this field gives the function the default namespace similar to the function registration. Must match the namespace of the function.  
```
I think my edits don't change the meaning, but please read my suggestion carefully and make sure I haven't changed this to say something you don't want it to!
